test/tuple.cpp

```diff
@@ -360,9 +360,53 @@ TEST(CampTuple, StructuredBindings)
 {
   auto t = camp::make_tuple(3, 9.9);
   auto [a, b] = t;
+  static_assert(std::is_same_v< decltype(a), int > );
+  static_assert(std::is_same_v< decltype(b), double > );
   ASSERT_EQ(a, 3);
   ASSERT_NEAR(b, 9.9, 1e-15);
 }
+
+TEST(CampTuple, StructuredBindingsConst)
+{
+  auto t = camp::make_tuple(3, 9.9);
+  const auto [a, b] = t;
+  static_assert(std::is_same_v< decltype(a), const int > );
+  static_assert(std::is_same_v< decltype(b), const double > );
+  ASSERT_EQ(a, 3);
+  ASSERT_NEAR(b, 9.9, 1e-15);
+}
+
+TEST(CampTuple, StructuredBindingsRef)
+{
+  auto t = camp::make_tuple(3, 9.9);
+  auto & [a, b] = t;
+  static_assert(std::is_same< decltype(a), int >::value ); // Not an int &
+  static_assert(std::is_same< decltype(b), double >::value ); // Not a double &
+  std::cout<<typeid(a).name()<<std::endl;
+  ASSERT_EQ(a, 3);
+  ASSERT_NEAR(b, 9.9, 1e-15);
+
+  a = 4; // Well...kind of an int &
+  b = 10.1; // also kind of a double &
+  ASSERT_EQ(camp::get<0>(t), 4);
+  ASSERT_NEAR(camp::get<1>(t), 10.1, 1e-15);
+}
+
+TEST(CampTuple, StructuredBindingsRefConst)
+{
+  auto t = camp::make_tuple(3, 9.9);
+  const auto & [a, b] = t;
+  static_assert(std::is_same< decltype(a), const int >::value );  // Not an const int &
+  static_assert(std::is_same< decltype(b), const double >::value ); // Not a const double &
+  std::cout<<typeid(a).name()<<std::endl;
+  ASSERT_EQ(a, 3);
+  ASSERT_NEAR(b, 9.9, 1e-15);
+
+  // these should fail to compile.
+  // a = 4;
+  // b = 10.1;
```

Comment:
@trws Do you have a preferred method for testing that compilations fail, or are you ok not testing? I don't think you can protect yourself fully given the `tuple` of `references` behavior?

Reply:
I have some infrastructure for compilation death tests, but pulled it out of mainline camp because checking the error messages proved rather brittle. If possible I'd say factor the test to use decltype on the binding to verify it's const-qualified as it should be, then we can keep the tests positive. If not I can dredge that stuff back up. It amounts to generating a non-auto cmake target that builds the test then registering cmake --build --target tgt as the test for ctest to run. Surprisingly robust, but also frustrating to make sure you're testing only what you want to test.

Reply:
Well...maybe the `static_assert( is_same<>)` checks are sufficient. It is up to you.

Reply:
I think the static_asserts are sufficient.